server/auth/types/openid/routes.ts

```diff
@@ -0,0 +1,213 @@
+            this.openIdAuthConfig.tokenEndpoint as string,
+            requestOptions
+          );
+          if (
+            !tokenResponse.res.statusCode ||
+            tokenResponse.res.statusCode < 200 ||
+            tokenResponse.res.statusCode > 299
+          ) {
+            return response.unauthorized();
+          }
+
+          const tokenPayload: any = this.parseTokenResponse(tokenResponse.payload as Buffer);
+
+          const idToken: string = tokenPayload.id_token;
+          const accessToken: string = tokenPayload.access_token;
+          const refreshToken: string = tokenPayload.refresh_token;
+          const expiresIn: number = tokenPayload.expires_in;
+
+          const user = await this.securityClient.authenticateWithHeader(
+            request,
+            this.openIdAuthConfig.authHeaderName as string,
+            `Bearer ${idToken}`
+          );
+
+          // set to cookie
+          const sessionStorage: SecuritySessionCookie = {
+            username: user.username,
+            credentials: {
+              authHeaderValue: `Bearer ${idToken}`,
+              refresh_token: refreshToken,
+              expires_at: Date.now() + expiresIn * 1000,
+            },
+            authType: 'openid',
+            expiryTime: Date.now() + this.config.cookie.ttl,
+          };
+          this.sessionStorageFactory.asScoped(request).set(sessionStorage);
+          return response.redirected({
+            headers: {
+              location: `${this.core.http.basePath.serverBasePath}/app/kibana`,
+            },
+          });
+        } catch (error) {
+          console.log(error); // TODO: change to logger
+          return response.unauthorized();
+        }
+      }
+    );
+
+    this.router.post(
+      {
+        path: `/auth/logout`,
+        validate: false,
+      },
+      async (context, request, response) => {
+        const cookie = await this.sessionStorageFactory.asScoped(request).get();
+        this.sessionStorageFactory.asScoped(request).clear();
+
+        const token = cookie?.credentials.authHeaderValue.split(' ')[1];
+        let requestQueryParameters = `?post_logout_redirect_uri=${this.getBaseRedirectUrl()}/app/kibana`;
+
+        let endSessionUrl = '/';
+        const customLogoutUrl = this.config.openid?.logout_url;
+        if (customLogoutUrl) {
+          endSessionUrl = customLogoutUrl + requestQueryParameters;
+        } else if (this.openIdAuthConfig.endSessionEndpoint) {
+          endSessionUrl =
+            this.openIdAuthConfig.endSessionEndpoint +
+            requestQueryParameters +
+            '&id_token_hint=' +
+            token;
+        }
+        return response.redirected({
+          headers: {
+            location: endSessionUrl,
+          }
+        });
+      }
+    );
+  }
+
+  private parseTokenResponse(payload: Buffer) {
+    const payloadString = payload.toString();
+    if (payloadString.trim()[0] === '{') {
```

Comment:
Just want to confirm whether there is any case where the return is non empty but without a leading "{" after trimming.

Reply:
if it is not JSON, then return the buffer content as parsed query string, see a couple lines below:
```
return parse(payloadString);
```

this follows in implementation of HAPI bell library at : https://github.com/hapijs/bell/blob/master/lib/oauth.js#L624